src/transformers/models/cogvlm/modeling_cogvlm.py

```diff
@@ -0,0 +1,1155 @@
+        pixel_values: List[List[torch.Tensor]] = None,
+        token_type_ids: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_values: Optional[List[torch.FloatTensor]] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, BaseModelOutputWithPast]:
+        r"""
+            labels (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
+                Labels for computing the language modeling loss. Indices should either be in `[0, ...,
+                config.vocab_size]` or -100 (see `input_ids` docstring). Tokens with indices set to `-100` are ignored,
+                the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`.
+
+        Returns:
+
+        Example:
+
+        ```python
+        >>> from transformers import CogvlmProcessor, CogvlmModel
+        >>> import torch
+        >>> import requests
+        >>> from PIL import Image
+
+        >>> processor = CogvlmProcessor.from_pretrained("THUDM/cogvlm-chat-hf")
+        >>> model = CogvlmModel.from_pretrained("THUDM/cogvlm-chat-hf")
+
+        >>> # load image
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+        >>> query = "Describe this image"
+
+        >>> prompt = f"Question: {query} Answer:"
+        >>> inputs = processor(images=image, text=prompt, return_tensors="pt")
+
+        >>> # forward pass
+        >>> with torch.no_grad():
+        ...     outputs = model(**inputs)
+
+        >>> last_hidden_state = outputs.last_hidden_state
+        ```
+        """
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if past_key_values is not None:
+            # generate mode with past_key_values. the image features are already mapped
+            # update attention_mask
+            if pixel_values is not None:
+                batch_size = pixel_values.shape[0]
+                vision_mask = torch.full(
+                    size=(batch_size, self.num_vision_tokens), fill_value=1, device=attention_mask.device
+                )
+                attention_mask = torch.cat(
+                    [attention_mask[:, :-1], vision_mask, attention_mask[:, -1].repeat(batch_size, 1)], dim=1
```

Comment:
We can slice the last attn mask in `prepare_inputs_for_generation` for consistency, since we do not use the rest